rfcs/0000-nixpkgs-breaking-change-policy.md

```diff
@@ -0,0 +1,75 @@
+- Merge to `staging`, notify dependents then eventually merge to `master`.
+- Gather a large number of fixes in the PR branch then merge to `master` once most or all dependencies are fixed.
+
+This aims to provide a uniform approach so that everyone involved knows what to expect and to allow further workflows including automation to be build on top of these procedures.
+
+## Past Discussions
+
+https://discourse.nixos.org/t/please-fix-all-packages-which-were-broken-by-the-qt-wrapping-changes/6444
+
+# Detailed design
+
+## Goals
+
+The following procedure will be used for making breaking changes. It has a few primary goals:
+
+- Avoid putting more burden than necessary on the dependency maintainer. If the maintainers of core derivations face toil proportionally to the number of transitive dependencies they will quickly become overloaded. These maintainers are arguably the most critical to nixpkgs and  their load needs to be kept manageable.
+- Avoid unnecessarily breaking packages for any period of time. There are a number of users on the `*-unstable` channels and it is annoying if packages in their configuration get broken.
+- Avoid breaking the `master` branch as much as possible. The `master` branch is used by the `*-unstable` channels and breaking this branch means that no updates can be pushed. This includes critical security updates as well as regular feature updates.
+
+## Procedure
+
+1. The maintainer will prepare a PR with their intended changes.
+2. The maintainer should test a sample of dependent derivations to ensure that their change is not unnecessarily or unintentionally breaking. (Example: Ensure that a critical output file was not forgotten) Note that sometimes it **is** necessary to break all dependent packages, and the maintainer is not required to avoid this.
+3. The maintainer will get the PR reviewed and approved.
+    1. It is **recommended but not required** to have some maintainers of dependent packages involved in the review to review if the breakage is justified.
+4. The maintainer will contact the maintainers of all dependent, broken packages (herein called sub-maintainers).
+    1. The sub-maintainers will be provided a link to the PR with the breaking changes as well as any context that will help them resolve the breakages.
+    2. The sub-maintainers are expected to update the derivations that they maintain within 7 days.
+        1. These changes should be merged to `master` following the regular nixpkgs PR process if they are backwards compatible. Otherwise they will be merged into the branch of the breaking PR.
+5. The maintainer must provide a grace period of at least 7 days from when the sub-maintainers were notified.
+6. The maintainer must merge `master` into their PR branch.
+7. The maintainer must mark all still-broken packages as [broken](https://nixos.org/manual/nixpkgs/stable/#sec-standard-meta-attributes).
+8. The maintainer can now merge to `master`.
```

Comment:
This assumes here that the change is sufficiently small it can go into master. With large changes that motivate this PR that is exactly not the case. The changes then go into staging, then staging-next and finally in master. That also gives an extra stabilization period on top of the proposed 7 days.

Reply:
Good point. I will update the target branch to be variable as it should still follow the correct procedures with that regard.

Reply:
For reference I believe this is the current policy for picking a target branch: https://github.com/NixOS/rfcs/blob/master/rfcs/0026-staging-workflow.md#detailed-design

Reply:
That's the RFC. The implementation is in the Nixpkgs manual, https://nixos.org/manual/nixpkgs/stable/#submitting-changes-commit-policy

Reply:
Ah, thanks!